internal/upstreamldap/upstreamldap_test.go

```diff
@@ -1916,3 +1959,243 @@ func TestGetDomainFromDistinguishedName(t *testing.T) {
 		})
 	}
 }
+
+func TestPwdUnchangedSinceLogin(t *testing.T) {
+	authTime := "2021-11-01T23:43:19.826433579Z" // this is the format that fosite automatically stores
+	authTimeParsed, err := time.Parse(time.RFC3339Nano, authTime)
+	require.NoError(t, err)
+	pwdResetTimeAfterAuthTime := "132803468800000000"  // Nov 2
+	pwdResetTimeBeforeAuthTime := "132801740800000000" // Oct 31
+	tests := []struct {
+		name       string
+		authTime   *time.Time
+		entry      *ldap.Entry
+		wantResult bool
+		wantErr    string
+	}{
+		{
+			name:     "happy path where password has not been reset since login",
+			authTime: &authTimeParsed,
+			entry: &ldap.Entry{
+				DN: "some-dn",
+				Attributes: []*ldap.EntryAttribute{
+					{
+						Name:   "pwdLastSet",
+						Values: []string{pwdResetTimeBeforeAuthTime},
+					},
+				},
+			},
+		},
+		{
+			name:     "password has been reset since login",
+			authTime: &authTimeParsed,
+			entry: &ldap.Entry{
+				DN: "some-dn",
+				Attributes: []*ldap.EntryAttribute{
+					{
+						Name:   "pwdLastSet",
+						Values: []string{pwdResetTimeAfterAuthTime},
+					},
+				},
+			},
+			wantErr: "password has changed since login. login time: 2021-11-01 23:43:19.826433579 +0000 UTC, password set time: 2021-11-02 17:14:40 +0000 UTC",
+		},
+		{
+			name:     "ldap timestamp is in the wrong format",
+			authTime: &authTimeParsed,
+			entry: &ldap.Entry{
+				DN: "some-dn",
+				Attributes: []*ldap.EntryAttribute{
+					{
+						Name:   "pwdLastSet",
+						Values: []string{"invalid"},
+					},
+				},
+			},
+			wantErr: "couldn't parse as timestamp",
+		},
+		{
+			name:     "no value for pwdLastSet attribute",
+			authTime: &authTimeParsed,
+			entry: &ldap.Entry{
+				DN:         "some-dn",
+				Attributes: []*ldap.EntryAttribute{},
+			},
+			wantErr: "expected to find 1 value for pwdLastSet attribute, but found 0",
+		},
+		{
+			name:     "too many values for pwdLastSet attribute",
+			authTime: &authTimeParsed,
+			entry: &ldap.Entry{
+				DN: "some-dn",
+				Attributes: []*ldap.EntryAttribute{
+					{
+						Name:   "pwdLastSet",
+						Values: []string{"val1", "val2"},
+					},
+				},
+			},
+			wantErr: "expected to find 1 value for pwdLastSet attribute, but found 2",
+		},
+	}
+	for _, test := range tests {
+		tt := test
+		t.Run(tt.name, func(t *testing.T) {
+			err := PwdUnchangedSinceLogin(tt.entry, provider2.StoredRefreshAttributes{AuthTime: *tt.authTime})
+			if tt.wantErr != "" {
+				require.Error(t, err)
+				require.Equal(t, tt.wantErr, err.Error())
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestWin32TimestampToTime(t *testing.T) {
```

Comment:
Add some perverse test cases such as:

1. min time
2. max time
3. (not sure if possible) times that cannot be precisely represented in `100s of ns`

Reply:
If we keep this code, we should add the max valid time test case.

Reply:
1. done
2. Unsure what you mean by max time, I have a test with a 21 digit number that isn't parseable to show that it is handled.
3. I could add this test, but it would just be another test case for ParseUint throwing an error when there's a non-integer value.

Reply:
Somehow I missed your follow up comment... I can add a max valid time.